arangod/GeneralServer/GeneralCommTask.cpp

```diff
@@ -81,11 +81,17 @@ void GeneralCommTask<T>::close(asio_ns::error_code const& ec) {
 template <SocketType T>
 void GeneralCommTask<T>::setTimeout(std::chrono::milliseconds millis) {
   _protocol->timer.expires_after(millis);
-  _protocol->timer.async_wait([self = CommTask::weak_from_this()](asio_ns::error_code ec) {
+  _protocol->timer.async_wait([self = CommTask::weak_from_this(), oldRequestCount = getRequestCount()](asio_ns::error_code ec) {
     std::shared_ptr<CommTask> s;
     if (ec || !(s = self.lock())) {  // was canceled / deallocated
       return;
     }
+
+    if (s->getRequestCount() != oldRequestCount) {
+      return;
+    }
+
+    s->setKeepAliveTimeoutReached();
```

Comment:
This looks very much like a data race since `_requestCount` and `_keepAliveTimeoutReached` are not atomic, but they are read/updated in different threads.

Reply:
Well, it does look like on! But this is all single threaded, all CommTasks are owned and handled by a single thread. (They are distributed across multiple io_contexts, each having a single thread)

Reply:
Both methods will only ever run on the single iothread of the iocontext. The only thing they need to ensure is that all goes well if both the idle timeout and the completion handler are scheduled basically at the same time. If the idle timeout is scheduled first, then the request must not be worked on, if the request handler comes first, the idle timeout must not be triggered any more. As far as I see, the proposed changes achieve exactly this.
I suggest to add a few more comments to explain this.

Reply:
Got it - each `IoContext` has only a single thread, so all operations triggered by that context are effectively serialized.